x-pack/apm-server/sampling/config.go

```diff
@@ -36,14 +36,16 @@ type LocalSamplingConfig struct {
 	// no greater than half of the TTL.
 	FlushInterval time.Duration
 
-	// MaxTraceGroups holds the maximum number of trace groups to track.
+	// MaxDynamicServices holds the maximum number of dynamic services to track.
 	//
-	// Once MaxTraceGroups is reached, any root transaction forming a new trace
-	// group will dropped.
-	MaxTraceGroups int
+	// Once MaxDynamicServices is reached, root transactions from a service that
+	// does not have an explicit policy defined may be dropped.
+	MaxDynamicServices int
 
-	// DefaultSampleRate is the default sample rate to assign to new trace groups.
-	DefaultSampleRate float64
+	// Policies holds local tail-sampling policies. Policies are matched in the
+	// order provided. Policies should therefore be ordered from most to least
```

Comment:
can/should we follow up with a task to order them ourselves from most to least specific, and not bother users with ordering?

Reply:
Great question!

I started out trying to make policies order-insensitive by defining an partial ordering (e.g. compare service name, then trace name, etc.).

In the end, particularly because everything is optional, I felt it would be too complex and subtle for users to understand how the policies would be applied, and it would be more straightforward to require explicit ordering. My expectation is that typical users will have a handful of policies, so ordering them would not be onerous.

Final point: we intend to develop a tail-sampling policy UI, which could take care of automatically ordering policies and could give some visual indications of precedence to reduce the subtlety.

In summary, I would like to wait and see.

Reply:
Fair enough :+1: